File: DataSheets/StateClassDataSheet.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core
Imports System.Reflection
Imports System.Globalization

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class StateClassDataSheet
    Inherits DataSheet

    Private m_SlxSheet As DataSheet
    Private m_SlySheet As DataSheet

    Protected Overrides Sub OnDataFeedsRefreshed()

        MyBase.OnDataFeedsRefreshed()

        Me.m_SlxSheet = Me.Project.GetDataSheet(DATASHEET_STATE_LABEL_X_NAME)
        Me.m_SlySheet = Me.Project.GetDataSheet(DATASHEET_STATE_LABEL_Y_NAME)

    End Sub

    Public Overrides Sub Validate(proposedRow As DataRow, transferMethod As DataTransferMethod)

        Dim ResetNull As Boolean = False

        'It's not possible to have NULL for a display member, but we want the user to be able to
        'have this name be auto-generated.  To avoid a validation failure because of a NULL here
        'we are going to fix up the Name with a default value until the validation is done and then
        'set it back to NULL.  The OnRowsAdded() override will auto-generate the name.  Note that
        'this should only be done for detached rows (i.e. rows that are being editing by a control
        'such as the DataGridView...)

        If (proposedRow(DATASHEET_NAME_COLUMN_NAME) Is DBNull.Value And
            proposedRow.RowState = DataRowState.Detached) Then

            proposedRow(DATASHEET_NAME_COLUMN_NAME) = "__SYNCROSIM_TEMP_VALUE__"
            ResetNull = True

        End If

        Try
            MyBase.Validate(proposedRow, transferMethod)
        Finally

            If (ResetNull) Then
                proposedRow(DATASHEET_NAME_COLUMN_NAME) = DBNull.Value
            End If

        End Try

    End Sub

    Protected Overrides Sub OnRowsAdded(sender As Object, e As DataSheetRowEventArgs)

        MyBase.OnRowsAdded(sender, e)

        Dim d As New Dictionary(Of String, Boolean)

        'Create a dictionary of the existing names

        For Each dr As DataRow In Me.GetData().Rows

            If (dr.RowState = DataRowState.Deleted) Then
                Continue For
            End If

            If (dr(DATASHEET_NAME_COLUMN_NAME) IsNot DBNull.Value) Then
                d.Add(CStr(dr(DATASHEET_NAME_COLUMN_NAME)), True)
            End If

        Next

        'If any name is NULL then add the name, checking for duplicates as we go

        For Each dr As DataRow In Me.GetData().Rows

            If (dr.RowState = DataRowState.Deleted) Then
                Continue For
            End If

            If (dr(DATASHEET_NAME_COLUMN_NAME) Is DBNull.Value) Then

                Dim slxdata As DataTable = Me.m_SlxSheet.GetData()
                Dim slydata As DataTable = Me.m_SlySheet.GetData()
                Dim slxid As Integer = CInt(dr(DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME))
                Dim slyid As Integer = CInt(dr(DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME))
                Dim slxname As String = CStr(DataTableUtilities.GetTableValue(slxdata, Me.m_SlxSheet.ValueMember, slxid, DATASHEET_NAME_COLUMN_NAME))
                Dim slyname As String = CStr(DataTableUtilities.GetTableValue(slydata, Me.m_SlySheet.ValueMember, slyid, DATASHEET_NAME_COLUMN_NAME))

                Dim n As String = GetNextName(slxname & ":" & slyname, d)
                dr(DATASHEET_NAME_COLUMN_NAME) = n

                d.Add(n, True)

            End If

        Next

    End Sub

    Protected Overrides Sub BeforeImportData(ByVal proposedData As DataTable)

        MyBase.BeforeImportData(proposedData)

        'We require the State Label X and Y values but they are not validated 
        'until after the call to BeforeImportData() so validate them now.

        For Each dr As DataRow In proposedData.Rows

            If (dr(DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME) Is DBNull.Value) Then

                ExceptionUtils.ThrowArgumentException("The data contains a NULL for '{0}'.",
                    Me.Columns(DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME).DisplayName)

            End If

            If (dr(DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME) Is DBNull.Value) Then

                ExceptionUtils.ThrowArgumentException("The data contains a NULL for '{0}'.",
                    Me.Columns(DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME).DisplayName)

            End If

        Next

        'Create a dictionary of the existing names

        Dim d As New Dictionary(Of String, Boolean)

        For Each dr As DataRow In proposedData.Rows

            If (dr(DATASHEET_NAME_COLUMN_NAME) IsNot DBNull.Value) Then
                d.Add(CStr(dr(DATASHEET_NAME_COLUMN_NAME)), True)
            End If

        Next

        'If any name is NULL then add the name, checking for duplicates as we go

        For Each dr As DataRow In proposedData.Rows

            If (dr(DATASHEET_NAME_COLUMN_NAME) Is DBNull.Value) Then

                Dim InitialName As String =
                    CStr(dr(DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME)) &
                    ":" &
                    CStr(dr(DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME))

                Dim FinalName As String = GetNextName(InitialName, d)
                dr(DATASHEET_NAME_COLUMN_NAME) = FinalName

                d.Add(FinalName, True)

            End If

        Next

    End Sub

    Private Shared Function GetNextName(
        ByVal proposedName As String,
        ByVal existingNames As Dictionary(Of String, Boolean)) As String

        Debug.Assert(Not String.IsNullOrEmpty(proposedName))
        Dim NewProposedName As String = proposedName

        Dim NextNum As Integer = 1
        Dim NextName As String = NewProposedName

        While (existingNames.ContainsKey(NextName))

            NextNum += 1
            NextName = String.Format(CultureInfo.InvariantCulture, "{0} ({1})", NewProposedName, NextNum)
            Debug.Assert(NextNum < 500)

        End While

        Return NextName

    End Function

End Class

File: DataSheets/StateClassDataSheet.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using SyncroSim.Core;
using System.Globalization;
using System.Collections.Generic;
using System.Diagnostics;

namespace SyncroSim.STSim
{
    internal class StateClassDataSheet : DataSheet
    {
        private DataSheet m_SlxSheet;
        private DataSheet m_SlySheet;

        protected override void OnDataFeedsRefreshed()
        {
            base.OnDataFeedsRefreshed();

            this.m_SlxSheet = this.Project.GetDataSheet(Strings.DATASHEET_STATE_LABEL_X_NAME);
            this.m_SlySheet = this.Project.GetDataSheet(Strings.DATASHEET_STATE_LABEL_Y_NAME);
        }

        public override void Validate(DataRow proposedRow, DataTransferMethod transferMethod)
        {
            bool ResetNull = false;

            //It's not possible to have NULL for a display member, but we want the user to be able to
            //have this name be auto-generated.  To avoid a validation failure because of a NULL here
            //we are going to fix up the Name with a default value until the validation is done and then
            //set it back to NULL.  The OnRowsAdded() override will auto-generate the name.  Note that
            //this should only be done for detached rows (i.e. rows that are being editing by a control
            //such as the DataGridView...)

            if (proposedRow[Strings.DATASHEET_NAME_COLUMN_NAME] == DBNull.Value && proposedRow.RowState == DataRowState.Detached)
            {
                proposedRow[Strings.DATASHEET_NAME_COLUMN_NAME] = "__SYNCROSIM_TEMP_VALUE__";
                ResetNull = true;
            }

            try
            {
                base.Validate(proposedRow, transferMethod);
            }
            finally
            {
                if (ResetNull)
                {
                    proposedRow[Strings.DATASHEET_NAME_COLUMN_NAME] = DBNull.Value;
                }
            }
        }

        protected override void OnRowsAdded(object sender, DataSheetRowEventArgs e)
        {
            base.OnRowsAdded(sender, e);

            Dictionary<string, bool> d = new Dictionary<string, bool>();

            //Create a dictionary of the existing names

            foreach (DataRow dr in this.GetData().Rows)
            {
                if (dr.RowState == DataRowState.Deleted)
                {
                    continue;
                }

                if (dr[Strings.DATASHEET_NAME_COLUMN_NAME] != DBNull.Value)
                {
                    d.Add(Convert.ToString(dr[Strings.DATASHEET_NAME_COLUMN_NAME], CultureInfo.InvariantCulture), true);
                }
            }

            //If any name is NULL then add the name, checking for duplicates as we go

            foreach (DataRow dr in this.GetData().Rows)
            {
                if (dr.RowState == DataRowState.Deleted)
                {
                    continue;
                }

                if (dr[Strings.DATASHEET_NAME_COLUMN_NAME] == DBNull.Value)
                {
                    DataTable slxdata = this.m_SlxSheet.GetData();
                    DataTable slydata = this.m_SlySheet.GetData();
                    int slxid = Convert.ToInt32(dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME], CultureInfo.InvariantCulture);
                    int slyid = Convert.ToInt32(dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME], CultureInfo.InvariantCulture);
                    string slxname = Convert.ToString(DataTableUtilities.GetTableValue(slxdata, this.m_SlxSheet.ValueMember, slxid, Strings.DATASHEET_NAME_COLUMN_NAME), CultureInfo.InvariantCulture);
                    string slyname = Convert.ToString(DataTableUtilities.GetTableValue(slydata, this.m_SlySheet.ValueMember, slyid, Strings.DATASHEET_NAME_COLUMN_NAME), CultureInfo.InvariantCulture);

                    string n = GetNextName(slxname + ":" + slyname, d);
                    dr[Strings.DATASHEET_NAME_COLUMN_NAME] = n;

                    d.Add(n, true);
                }
            }
        }

        protected override void BeforeImportData(DataTable proposedData)
        {
            base.BeforeImportData(proposedData);

            //We require the State Label X and Y values but they are not validated 
            //until after the call to BeforeImportData() so validate them now.

            foreach (DataRow dr in proposedData.Rows)
            {
                if (dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME] == DBNull.Value)
                {
                    ExceptionUtils.ThrowArgumentException(
                        "The data contains a NULL for '{0}'.", 
                        this.Columns[Strings.DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME].DisplayName);
                }

                if (dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME] == DBNull.Value)
                {
                    ExceptionUtils.ThrowArgumentException(
                        "The data contains a NULL for '{0}'.", 
                        this.Columns[Strings.DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME].DisplayName);
                }
            }

            //Create a dictionary of the existing names

            Dictionary<string, bool> d = new Dictionary<string, bool>();

            foreach (DataRow dr in proposedData.Rows)
            {
                if (dr[Strings.DATASHEET_NAME_COLUMN_NAME] != DBNull.Value)
                {
                    d.Add(Convert.ToString(dr[Strings.DATASHEET_NAME_COLUMN_NAME], CultureInfo.InvariantCulture), true);
                }
            }

            //If any name is NULL then add the name, checking for duplicates as we go

            foreach (DataRow dr in proposedData.Rows)
            {
                if (dr[Strings.DATASHEET_NAME_COLUMN_NAME] == DBNull.Value)
                {
                    string InitialName = Convert.ToString(
                        dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME], CultureInfo.InvariantCulture) + 
                        ":" + 
                        Convert.ToString(dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME], CultureInfo.InvariantCulture);

                    string FinalName = GetNextName(InitialName, d);
                    dr[Strings.DATASHEET_NAME_COLUMN_NAME] = FinalName;

                    d.Add(FinalName, true);
                }
            }
        }

        private static string GetNextName(string proposedName, Dictionary<string, bool> existingNames)
        {
            Debug.Assert(!string.IsNullOrEmpty(proposedName));
            string NewProposedName = proposedName;

            int NextNum = 1;
            string NextName = NewProposedName;

            while (existingNames.ContainsKey(NextName))
            {
                NextNum += 1;
                NextName = string.Format(CultureInfo.InvariantCulture, "{0} ({1})", NewProposedName, NextNum);
                Debug.Assert(NextNum < 500);
            }

            return NextName;
        }
    }
}
